cmd/clusterctl/pkg/client/config/providers_client.go

```diff
@@ -0,0 +1,203 @@
+	Get(name string) (*Provider, error)
+}
+
+// providersClient implements ProvidersClient.
+type providersClient struct {
+	reader Reader
+}
+
+// ensure providersClient implements ProvidersClient.
+var _ ProvidersClient = &providersClient{}
+
+func newProvidersClient(reader Reader) *providersClient {
+	return &providersClient{
+		reader: reader,
+	}
+}
+
+func (p *providersClient) defaults() []Provider {
+
+	// clusterctl includes an hard-coded list of "default" providers in order to provide to the users the simplest
+	// out-of-box experience. Other providers can be added by using the clusterctl configuration file.
+
+	// if you are a developer of a community maintained provider, you can send a PR to extend the following list.
+
+	defaults := []Provider{
+		// cluster API core provider
+		&provider{
+			name:         CAPI,
+			url:          "https://github.com/kubernetes-sigs/cluster-api/releases/latest/cluster-api-components.yaml",
+			providerType: clusterctlv1.CoreProviderType,
+		},
+
+		// Infrastructure providersClient
+		&provider{
+			name:         "aws",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/latest/infrastructure-components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:         "docker",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-docker/releases/latest/provider_components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:         "vsphere",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/releases/latest/infrastructure-components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+
+		// Bootstrap providersClient
+		// TODO: CABPK in v1alpha3 will be included into CAPI, so this entry can be removed as soon as v1alpha3 is ready for test
+		&provider{
+			name:         "kubeadm",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases/latest/bootstrap-components.yaml",
+			providerType: clusterctlv1.BootstrapProviderType,
+		},
+	}
+
+	// ensure defaults are consistently sorted
+	sort.Slice(defaults, func(i, j int) bool {
+		return defaults[i].Name() < defaults[j].Name()
+	})
+
+	return defaults
+}
+
+func (p *providersClient) List() ([]Provider, error) {
+	// Creates a maps with all the defaults provider configurations
+	rMap := map[string]Provider{}
+	for _, r := range p.defaults() {
+		rMap[r.Name()] = r
+	}
+
+	// Gets user defined provider configurations, validate them, and merges with
+	// hard-coded configurations handling conflicts (user defined take precedence on hard-coded)
+	type configProvider struct {
+		Name string                    `json:"name,omitempty"`
+		URL  string                    `json:"url,omitempty"`
+		Type clusterctlv1.ProviderType `json:"type,omitempty"`
+	}
+
+	var ur []configProvider //nolint
```

Comment:
What's `ur`? Let's expand this variable?

Reply:
done